play-ws-standalone/src/main/scala/play/api/libs/ws/StandaloneWSRequest.scala

```diff
@@ -13,14 +13,7 @@ import scala.concurrent.duration.Duration
  * A WS Request builder.
  */
 trait StandaloneWSRequest {
-  // There is an issue with type members, because when you stack `Self` references,
-  // they don't have a connection to each other that ties them back to the parent
-  // `Self` reference.  The way around this is to do F-bounded polymorphism directly
-  // on the type member, which roots it to the parent and doesn't mess up the type
-  // signature the way that a self-recursive type parameter would.  The implementation
-  // will always bind this to a relevant type, so we only have to make the compiler
-  // happy here.
-  type Self <: StandaloneWSRequest { type Self <: StandaloneWSRequest.this.Self }
+  type Self <: StandaloneWSRequest
```

Comment:
The f-bounds can be preserved by writing
```suggestion
  type Self <: StandaloneWSRequest { type Self = StandaloneWSRequest.this.Self }
```
cf https://github.com/lampepfl/dotty/issues/12731#issuecomment-909666437

Reply:
```
play-ahc-ws-standalone/src/main/scala/play/api/libs/ws/ahc/StandaloneAhcWSRequest.scala:51:17: incompatible type in overriding
[error] type Self <: play.api.libs.ws.StandaloneWSRequest{type Self = StandaloneAhcWSRequest.this.Self} (defined in trait StandaloneWSRequest);
[error]  found   : play.api.libs.ws.StandaloneWSRequest
[error]  required:  <: play.api.libs.ws.StandaloneWSRequest{type Self = StandaloneAhcWSRequest.this.Self}
[error]     (which expands to)   <: play.api.libs.ws.StandaloneWSRequest{type Self = play.api.libs.ws.StandaloneWSRequest}
[error]   override type Self     = StandaloneWSRequest
[error]                 ^
```

https://github.com/playframework/play-ws/blob/66725c3d900cfb806b7c52e1ea1808a8a729ba36/play-ahc-ws-standalone/src/main/scala/play/api/libs/ws/ahc/StandaloneAhcWSRequest.scala#L51

Reply:
That line looks like it's explicitly bypassing the point of having a precise Self type by returning a less precise type instead, seems like conflicting requirements.

Reply:
```scala
trait StandaloneWSRequest { self =>
  type Self <: StandaloneWSRequest { type Self = self.Self }
```

Reply:
> trait StandaloneWSRequest { self =>
>   type Self <: StandaloneWSRequest { type Self = self.Self }
> or
> 
> trait StandaloneWSRequest { self =>
>   type Self <: StandaloneWSRequest { type Self <: self.Self }

compile error 😢 


```
[error] type Self <: play.api.libs.ws.StandaloneWSRequest{type Self = StandaloneAhcWSRequest.this.Self} (defined in trait StandaloneWSRequest);
[error]  found   : play.api.libs.ws.StandaloneWSRequest
[error]  required:  <: play.api.libs.ws.StandaloneWSRequest{type Self = StandaloneAhcWSRequest.this.Self}
[error]     (which expands to)   <: play.api.libs.ws.StandaloneWSRequest{type Self = play.api.libs.ws.StandaloneWSRequest}
[error]   override type Self     = StandaloneWSRequest
[error]                 ^
```

```
play-ahc-ws-standalone/src/test/scala/play/api/libs/ws/ahc/AhcWSRequestSpec.scala:663:22 
[error] 663 |      .withHttpHeaders("Content-Length" -> "9001") // add a meaningless content length here...
[error]     |                      ^
[error]     |Recursion limit exceeded.
[error]     |Maybe there is an illegal cyclic reference?
[error]     |If that's not the case, you could also try to increase the stacksize using the -Xss JVM option.
[error]     |A recurring operation is (inner to outer):
[error]     |
[error]     |  subtype ... (caught cyclic reference) ... <:< ... (caught cyclic reference) ...
[error]     |  subtype ... (caught cyclic reference) ... <:< ... (caught cyclic reference) ...
[error]     |  subtype ... (caught cyclic reference) ... <:< ... (caught cyclic reference) ...
[error]     |  subtype ... (caught cyclic reference) ... <:< ... (caught cyclic reference) ...
[error]     |  subtype ... (caught cyclic reference) ... <:< ... (caught cyclic reference) ...
[error]     |  subtype ... (caught cyclic reference) ... <:< ... (caught cyclic reference) ...
[error]     |  subtype ... (caught cyclic reference) ... <:< ... (caught cyclic reference) ...
[error]     |  subtype ... (caught cyclic reference) ... <:< ... (caught cyclic reference) ...
[error]     |  subtype ... (caught cyclic reference) ... <:< ... (caught cyclic reference) ...
[error]     |  subtype ... (caught cyclic reference) ... <:< ... (caught cyclic reference) ...
[error]     |  ...
[error]     |
[error]     |  subtype play.api.libs.ws.StandaloneWSRequest#Self#Self <:< play.api.libs.ws.StandaloneWSRequest{
[error]     |  Self <: 
[error]     |    play.api.libs.ws.StandaloneWSRequest{
[error]     |      Self <: play.api.libs.ws.StandaloneWSRequest#Self#Self
[error]     |    }#Self
[error]     |}
[error]     |  subtype play.api.libs.ws.StandaloneWSRequest{
[error]     |  Self <: play.api.libs.ws.StandaloneWSRequest#Self#Self
[error]     |} & play.api.libs.ws.StandaloneWSRequest#Self <:< play.api.libs.ws.StandaloneWSRequest{
[error]     |  Self <: 
[error]     |    play.api.libs.ws.StandaloneWSRequest{
[error]     |      Self <: play.api.libs.ws.StandaloneWSRequest#Self#Self
[error]     |    }#Self
[error]     |}
[error]     |  subtype play.api.libs.ws.StandaloneWSRequest#Self#Self <:< play.api.libs.ws.StandaloneWSRequest{
[error]     |  Self <: 
[error]     |    play.api.libs.ws.StandaloneWSRequest{
[error]     |      Self <: play.api.libs.ws.StandaloneWSRequest#Self#Self
[error]     |    }#Self
[error]     |}
[error]     |  subtype play.api.libs.ws.StandaloneWSRequest{
[error]     |  Self <: play.api.libs.ws.StandaloneWSRequest#Self#Self
[error]     |} & play.api.libs.ws.StandaloneWSRequest#Self <:< play.api.libs.ws.StandaloneWSRequest{
[error]     |  Self <: 
[error]     |    play.api.libs.ws.StandaloneWSRequest{
[error]     |      Self <: play.api.libs.ws.StandaloneWSRequest#Self#Self
[error]     |    }#Self
[error]     |}
[error]     |  subtype play.api.libs.ws.StandaloneWSRequest#Self#Self <:< play.api.libs.ws.StandaloneWSRequest{
[error]     |  Self <: 
[error]     |    play.api.libs.ws.StandaloneWSRequest{
[error]     |      Self <: play.api.libs.ws.StandaloneWSRequest#Self#Self
[error]     |    }#Self
[error]     |}
[error]     |  subtype play.api.libs.ws.StandaloneWSRequest{
[error]     |  Self <: play.api.libs.ws.StandaloneWSRequest#Self#Self
[error]     |} & play.api.libs.ws.StandaloneWSRequest#Self <:< play.api.libs.ws.StandaloneWSRequest{
[error]     |  Self <: 
[error]     |    play.api.libs.ws.StandaloneWSRequest{
[error]     |      Self <: play.api.libs.ws.StandaloneWSRequest#Self#Self
[error]     |    }#Self
[error]     |}
[error]     |  subtype play.api.libs.ws.StandaloneWSRequest#Self#Self <:< play.api.libs.ws.StandaloneWSRequest{
[error]     |  Self <: 
[error]     |    play.api.libs.ws.StandaloneWSRequest{
[error]     |      Self <: play.api.libs.ws.StandaloneWSRequest#Self#Self
[error]     |    }#Self
[error]     |}
[error]     |  subtype play.api.libs.ws.StandaloneWSRequest{
[error]     |  Self <: play.api.libs.ws.StandaloneWSRequest#Self#Self
[error]     |} & play.api.libs.ws.StandaloneWSRequest#Self <:< ... (caught cyclic reference) ...
[error]     |  subtype play.api.libs.ws.StandaloneWSRequest#Self#Self <:< ... (caught cyclic reference) ...
[error]     |  subtype play.api.libs.ws.StandaloneWSRequest#Self#Self <:< ?{ sign: ? }
```

Reply:
Compile without error for me

Reply:
- https://github.com/xuwei-k/play-ws/commit/0ed9f8d5b50c084c75e76e578a1d7ea825829e0b
- https://github.com/xuwei-k/play-ws/commit/2e4e6afff51a6afbb46c0d9cc3162595b047d2d0
- https://github.com/xuwei-k/play-ws/runs/4772742406#step:5:63
- https://github.com/xuwei-k/play-ws/runs/4772746164